### PR TITLE
Add commit best practices

### PR DESCRIPTION
**Pull request by slaperche-scality**
**Component**:

docs

**Context**: 

We want to improve the long term maintainability, and having goog Git history is a important part of it

**Summary**:

Define commit best practices.

**Acceptance criteria**:

Everyone's approval

---

Closes: #1361

### CODE REVIEW

**Review by ycointe — DISMISSED** [review]
LGTM !

File: docs/developer/best_practices/commit.rst [hunk 1]
```diff
@@ -0,0 +1,357 @@
+Commit best practices
+---------------------
+
+How to split a changes into commits
```
Comment (charlesprost) [inline docs/developer/best_practices/commit.rst]:
```suggestion
How to split a change into commits
```

File: docs/developer/best_practices/commit.rst [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+^^^^^^^^
+
+Cosmetic changes
+::::::::::::::::
+
+Use a dedicated commit when you want to make cosmetic changes to the code
+(linting, whitespaces, alignment, renaming, …).
+
+Mixing cosmetics and functional changes is bad because the cosmetics (which
+tends to generate a lot of diff/noise) will obscure the important functional
```
Comment (charlesprost) [inline docs/developer/best_practices/commit.rst]:
```suggestion
tend to generate a lot of diff/noise) will obscure the important functional
```

File: docs/developer/best_practices/commit.rst [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+as well.
+
+Mixing unrelated changes
+::::::::::::::::::::::::
+
+It is sometimes tempting to do small unrelated changes as you are working on
+something else in the same code area.
+Please refrain to do so, or at least do it in a dedicated commit.
+
+Mixing non-related changes into the same commit makes revert and cherry-pick
```
Comment (charlesprost) [inline docs/developer/best_practices/commit.rst:74]: And understanding as well!

File: docs/developer/best_practices/commit.rst [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+
+- you need access to an external service to get the whole context, which goes
+  against one of biggest advantage of the distributed SCM (having all the
+  information you need offline, from your local copy of the repository).
+
+Anatomy of a good commit message
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+A commit is composed of a subject, a body and a footer. A blank line separates
+the subject from body.
```
Comment (charlesprost) [inline docs/developer/best_practices/commit.rst]: And what separates the footer from the body?

File: docs/developer/best_practices/commit.rst [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+    detecting macOS and forcing the use of threads in that case.
+
+    [1]: http://sealiesoftware.com/blog/archive/2017/6/5/Objective-C_and_fork_in_macOS_1013.html
+    [2]: https://blog.phusion.nl/2017/10/13/why-ruby-app-servers-break-on-macos-high-sierra-and-what-can-be-done-about-it/
+
+    Closes: #1354
+
+Some things to note about this commit message:
+
+- Observed problem is described: parallel builds crash on MacOS
```
Comment (charlesprost) [inline docs/developer/best_practices/commit.rst]:
```suggestion
- Observed problem is described: parallel builds crash on MacOS.
```

File: docs/developer/best_practices/commit.rst [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+
+- Root cause is analyzed: OS security measure + thread/fork mix
```
Comment (charlesprost) [inline docs/developer/best_practices/commit.rst]:
```suggestion
- Root cause is analyzed: OS security measure + thread/fork mix.
```

**Review by charlesprost — DISMISSED** [review]

File: docs/developer/best_practices/commit.rst (hunk 1 above)
Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst]: Done

File: docs/developer/best_practices/commit.rst (hunk 2 above)
Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst]: Done

File: docs/developer/best_practices/commit.rst (hunk 4 above)
Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst]: A blank line as well, I'll add it.

File: docs/developer/best_practices/commit.rst (hunk 3 above)
Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst:74]: I'll add it, good point :slightly_smiling_face:

File: docs/developer/best_practices/commit.rst (hunk 5 above)
Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst]: Done

File: docs/developer/best_practices/commit.rst (hunk 6 above)
Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst]: Done

File: docs/developer/best_practices/commit.rst [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+part of the code documentation and makes them helpful for other developers to
+understand your code.
+
+Last but not least: commit messages can also be used for automating tasks such
+as issue management and changelog generation.
+
+Note that it is important to have all the necessary information in the commit
+message, instead of having them (only) in the related issue, because:
+
+- issues can contains troubleshooting/design discussion/investigation with a
```
Comment (rdebay-scality) [inline docs/developer/best_practices/commit.rst]: "issue can contain" ?

Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst]: Done.

**Review by rdebay-scality — DISMISSED** [review]

File: docs/developer/best_practices/commit.rst [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+
+When reviewing a change, do not simply look at the correctness of the code:
+review the commit message itself and request improvements to its content.
+Look out for commits that can be divided, ensure that cosmetic changes are not
+mixed with functional changes, …
+
+The goal here is to improve the long term maintainability, by a wide variety of
+developers. Considering that we are an Open Source project, external
+contributors may only have the Git history to get some context so it is
+important to have a useful Git history.
```
Comment (TeddyAndrieux) [inline docs/developer/best_practices/commit.rst]: This doc should be used by open source contributors as well. IMO we should not talk about the fact that we are an "Open Source project" in the doc. 
I would suggest something 
```suggestion
The goal here is to improve the long term maintainability, by a wide variety of
developers who may only have the Git history to get some context so it is
important to have a useful Git history.
```

(Suggestion does not work very well with multi line comment :stuck_out_tongue: )

Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst]: Fair enough. Done!

File: docs/developer/best_practices/commit.rst [hunk 9]
```diff
@@ -0,0 +1,356 @@
+Commit best practices
```
Comment (NicolasT) [inline docs/developer/best_practices/commit.rst]: Shouldn't we use *titlecase*?

```suggestion
Commit Best Practices
```

File: docs/developer/best_practices/commit.rst [hunk 10]
```diff
@@ -0,0 +1,356 @@
...
+---------------------
+
+How to split a change into commits
```
Comment (NicolasT) [inline docs/developer/best_practices/commit.rst:4]: The *why* should come before the *how*.

File: docs/developer/best_practices/commit.rst [hunk 11]
```diff
@@ -0,0 +1,356 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The golden rule to create good commits is to ensure that there is only one
+logical change per commit.
```
Comment (NicolasT) [inline docs/developer/best_practices/commit.rst]: I'd rephrase this, because *logical* is quite broad and non-precise.

IMO, the content of a commit should be `a reversible unit`, i.e., something which, when `git revert`ed, doesn't break the product. Related, but less important (and likely more related to what you describe as `logical unit`) is something which can be `git cherry-pick`ed.

Comment (NicolasT) [reply docs/developer/best_practices/commit.rst]: Looks like you refer to this before. However, I still think *logical* is too imprecise.

File: docs/developer/best_practices/commit.rst [hunk 12]
```diff
@@ -0,0 +1,356 @@
...
+
+- make ``git log`` and ``git blame`` way more useful
+
+Examples
+^^^^^^^^
+
+Cosmetic changes
+::::::::::::::::
+
+Use a dedicated commit when you want to make cosmetic changes to the code
```
Comment (NicolasT) [inline docs/developer/best_practices/commit.rst:33]: *unless these cosmetic changes change code that's introduced in the same PR*

File: docs/developer/best_practices/commit.rst [hunk 13]
```diff
@@ -0,0 +1,356 @@
...
+(linting, whitespaces, alignment, renaming, …).
+
+Mixing cosmetics and functional changes is bad because the cosmetics (which
+tend to generate a lot of diff/noise) will obscure the important functional
+changes, making it harder to correctly determine whether the change is correct
+during the review.
```
Comment (NicolasT) [inline docs/developer/best_practices/commit.rst:39]: *Ideally, purely cosmetic changes which inflate the number of changes in a PR significantly, should go in a separate PR*

File: docs/developer/best_practices/commit.rst [hunk 14]
```diff
@@ -0,0 +1,356 @@
...
+- one commit for the functional changes: `3367fabdefc0b35d34bf7cf2fb0d33ff81f9fd5a <https://github.com/scality/metalk8s/commit/3367fabdefc0b35d34bf7cf2fb0d33ff81f9fd5a>`_
+
+Refactoring
+:::::::::::
+
+When introducing new features, you often have to add new helpers or refactor
+existing code. In such case, instead of having single commit with everything
+inside, you can either:
+
+1. first add a new helper: `29f49cbe9dfa <https://github.com/scality/metalk8s/commit/29f49cbe9dfa0b824c818d25d4a2f6965351e65d>`_
```
Comment (NicolasT) [inline docs/developer/best_practices/commit.rst:57]: I disagree with this one and don't consider it a 'good' commit. It adds utility code which is not used anywhere.

This breaks the `git revert` rule.

Instead, the commit adding the utility function could've also updated code to *use* this utility, if any. If no existing code can use it, then introduce it *with* the 'feature' code that *will*, otherwise the `git revert`/`git cherry-pick` rule doesn't apply again.

File: docs/developer/best_practices/commit.rst [hunk 15]
```diff
@@ -0,0 +1,356 @@
...
+
+2. then use it in new code: `7e47310a8f20 <https://github.com/scality/metalk8s/commit/7e47310a8f20fd49f0ad36707b20e6c2a53df638>`_
+
+Or:
+
+1. first add the new code: `5b2a6d5fa498 <https://github.com/scality/metalk8s/commit/5b2a6d5fa49815180a2effdd37cb58542e83b5a5>`_
```
Comment (NicolasT) [inline docs/developer/best_practices/commit.rst:63]: Again, I'm not sure doing this in 2 commits brings any value. It'd require *both* commits to be reverted and/or cherry-picked to 'work'.

File: docs/developer/best_practices/commit.rst [hunk 16]
```diff
@@ -0,0 +1,356 @@
...
+
+2. then refactor the now duplicated code: `ac08d0f53a83 <https://github.com/scality/metalk8s/commit/ac08d0f53a835a0b2bc61c1fe5b7317bf4d6550c>`_
+
+If you add tests (and you should), they can be introduced in a dedicated commit
```
Comment (NicolasT) [inline docs/developer/best_practices/commit.rst]: No. This relates to `git bisect`. Say, the code is wrong. Then, the test fail. Now, if you `git bisect`, you'd *find* the commit introducing the *test*, not the one which contains the 'bad code'.

So, no. Code and tests for said code go in one commit.

File: docs/developer/best_practices/commit.rst [hunk 17]
```diff
@@ -0,0 +1,356 @@
...
+for each line, the identifier of the last commit that changed the line.
+
+Unlike a comment in the code (which applies to a single line or file), a commit
+message applies to a logical change and thus can provide information on the
+design of the code and why the change was done. This makes commit messages a
+part of the code documentation and makes them helpful for other developers to
+understand your code.
+
+Last but not least: commit messages can also be used for automating tasks such
+as issue management and changelog generation.
```
Comment (NicolasT) [inline docs/developer/best_practices/commit.rst]: We don't want to use commit messages for changelog generation because they're immutable (among other reasons).

See https://docs.google.com/document/d/1NUUShY9VBdLa3s0ePR5swptBtszXwM9qvFsF52L22L4/edit?usp=sharing

File: docs/developer/best_practices/commit.rst [hunk 18]
```diff
@@ -0,0 +1,356 @@
...
+harder (and understanding as well).
+
+The pull request `#1846 <https://github.com/scality/metalk8s/pull/1846>`_ is a good example. It tackles three issues at once: `#1830 <https://github.com/scality/metalk8s/issues/1830>`_
+and `#1831 <https://github.com/scality/metalk8s/issues/1831>`_ (because they are similar) and `#839 <https://github.com/scality/metalk8s/issues/839>`_ (because it was making the other
+changes easier), but it uses distincts commits for each issue.
+
+How to write a commit message
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Why do we need commit message
```
Comment (NicolasT) [inline docs/developer/best_practices/commit.rst]:
```suggestion
Why do we need commit messages
```

File: docs/developer/best_practices/commit.rst [hunk 19]
```diff
@@ -0,0 +1,356 @@
...
+Note that it is important to have all the necessary information in the commit
+message, instead of having them (only) in the related issue, because:
+
+- the issue can contain troubleshooting/design discussion/investigation with a
+  lot of back and forth, which makes hard to get the gist of it.
+
+- you need access to an external service to get the whole context, which goes
+  against one of biggest advantage of the distributed SCM (having all the
+  information you need offline, from your local copy of the repository).
+
```
Comment (NicolasT) [inline docs/developer/best_practices/commit.rst:109]: Related to the above, switching tracking system may lose history. Switching VCS should not.

File: docs/developer/best_practices/commit.rst [hunk 20]
```diff
@@ -0,0 +1,356 @@
...
+It is composed of:
+
+- a topic, usually the name of the affected component (``ui``, ``build``,
+  ``docs``, …)
+
+- a slash and then the name of the sub-component (optional)
+
+- a colon
+
+- the description of the change
```
Comment (NicolasT) [inline docs/developer/best_practices/commit.rst:150]: - no capital

File: docs/developer/best_practices/commit.rst [hunk 21]
```diff
@@ -0,0 +1,356 @@
...
+
+- ``build/package: factorize task_dep in DEBPackage``
+
+- ``ui/volume: add banner when failed to create volume``
+
+If several components are affected:
+
+- split your commit (preferred)
+
+- pick only the most affected one
```
Comment (NicolasT) [inline docs/developer/best_practices/commit.rst:164]: or comma-separate?

File: docs/developer/best_practices/commit.rst [hunk 22]
```diff
@@ -0,0 +1,356 @@
...
+
+Example:
+
+.. code:: text
+
+    topic: description
+
+    [… commit message body …]
+
+    Refs: #XXXXX, #YYYYY
```
Comment (NicolasT) [inline docs/developer/best_practices/commit.rst]: Prefer one per line.
Also, I tend to do provide not only `See: #XXXX`, but also `See: https://github.com/scality/metalk8s/issues/XXXX` or whatever, such that one can easily click-to-open in a terminal.

File: docs/developer/best_practices/commit.rst [hunk 23]
```diff
@@ -0,0 +1,356 @@
...
+Commit `b531290c04c4 <https://github.com/scality/metalk8s/commit/b531290c04c45a01cd103a85431e2428b98d340e>`_
+''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
+
+.. code:: text
+
+    Add gzip to nginx conf
+
+    This will decrease the size of the file the client need to download
+    In the current version we have ~7x improvement.
+    From 3.17Mb to 0.470Mb send to the client
```
Comment (NicolasT) [inline docs/developer/best_practices/commit.rst:269]: And this is exactly why we can't generate changelogs from the commits ;-)

**Review by TeddyAndrieux — DISMISSED** [review]
LGTM

**Review by NicolasT — DISMISSED** [review]

File: docs/developer/best_practices/commit.rst (hunk 9 above)
Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst]: Done

File: docs/developer/best_practices/commit.rst (hunk 15 above)
Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst:63]: True for the revert/cherry-pick part.

But it does make the review easier: you can review the feature and the refactor separately, while conflating both may be harder to review.

File: docs/developer/best_practices/commit.rst (hunk 12 above)
Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst:33]: If it's code introduced in this PR, then it's new code and you will write it "correctly" when you introduce it.

Or do you talk about a commit that changes some existing code, and thus reformat it as it modifies it?
Yeah, in this case you won't do two commits (one which update the code but keep the bad formatting and then one to fix the formatting), you'll edit the code and fix the formatting directly).

File: docs/developer/best_practices/commit.rst (hunk 13 above)
Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst:39]: Indeed, I'll add it.

File: docs/developer/best_practices/commit.rst (hunk 14 above)
Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst:57]: The function to check if an image exists is not tied to the `uptodate` one.

I mean, yes `uptodate` does need this helper (and thus if you want to cherry-pick this "feature" you'll have to cherry-pick the helper as well), but the helper could also be used in another independent context (pre-check that some images exists, …) and thus could be cherry-picked independently.
Likewise, you may want to revert the `uptodate` stuff while keeping the helper around for another use.

File: docs/developer/best_practices/commit.rst (hunk 16 above)
Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst]: Hum, good point indeed!

I was thinking  how we sometimes wrote test today (especially E2E one):  write a huge chunk of code conflating required steps and the test scenario.
Adding this to the feature would make for a huge commit.

But I guess if we wrote the steps and everything needed for the tests in prior commits, and then have a commit with the feature AND only the test scenario, that would be doable.

I'll remove this sentence then :slightly_smiling_face: 

(now, to be honest, I doubt that `bisect` works well on a codebase where the CI doesn't enforce that the code at least "build" for each commit :stuck_out_tongue: )

File: docs/developer/best_practices/commit.rst (hunk 18 above)
Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst]: Done

File: docs/developer/best_practices/commit.rst (hunk 10 above)
Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst:4]: Done (I think)

File: docs/developer/best_practices/commit.rst (hunk 17 above)
Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst]: Yeah, @gdemonet shared this document with me afterward.
That being said, in general it's a possible use case of commit messages.

But yeah, since it's doesn't fit our needs + those guideline are specific to this project, mentioning it here is irrelevant, so I'll remove the changelog example.

File: docs/developer/best_practices/commit.rst (hunk 21 above)
Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst:164]: It's an option, but it will nibble at the already tight character budget of the subject line.

So I'm not sure it's a good one :/

File: docs/developer/best_practices/commit.rst (hunk 22 above)
Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst]: Ok for one per line, I'll update the example.

Good idea for also providing a real hyperlink!

File: docs/developer/best_practices/commit.rst (hunk 23 above)
Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst:269]: Indeed ^^

File: docs/developer/best_practices/commit.rst (hunk 20 above)
Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst:150]: :+1: 

And since subject line are more like a title, they shouldn't end with a period (I'll remove that part).

File: docs/developer/best_practices/commit.rst (hunk 19 above)
Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst:109]: Yup, I forgot to mention that one. I'll add it.

File: docs/developer/best_practices/commit.rst (hunk 11 above)
Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst]: Yeah, `logical` is kinda unclear I agree.

I'll rewrite this part using a definition more similar to your yours.

**Review by ChengYanJin — DISMISSED** [review]

**Review by sayf-eddine-scality — DISMISSED** [review]

File: docs/developer/best_practices/commit.rst (hunk 15 above)
Comment (carlito-scality) [reply docs/developer/best_practices/commit.rst:63]: Personally, it is easier for me to review if they are in one commit since I would like know why is it refactored and how is it used

Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst:63]: >  I would like know why is it refactored and how is it used

But you'll know why it is refactored (because the previous commit implementing the feature introduced code duplication) and you'll see how it's used.

Having two commit also have the following advantage:
- first you implement the new feature, testing it manually/adding unit test
- then you refactor and you know that it works because you have the unit test/manual test from the previous commit.

Now, and maybe that's not clear here, the goal of those examples is to help people who have no idea how they could split things.
These aren't hard and fast rules to be applied in all cases, I trust people to follow their judgement here.

**Review by jbertran — DISMISSED** [review]

File: docs/developer/best_practices/commit.rst (hunk 15 above)
Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst:63]: Now, if this example is really controversial/confusing I can remove it.

It's no big deal, given that:
- people usually split their changes into commit already (most of the time at least), our current biggest problem are the commit messages…
- those examples are only a guidance at best, not something to strictly enforce.

Opinion @NicolasT , @carlito-scality ?

File: docs/developer/best_practices/commit.rst [hunk 24]
```diff
@@ -0,0 +1,372 @@
+Commit Best Practices
+---------------------
+
+How to split a change into commits
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Why do we need to split changes into commits
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+This has several advantages amongst whose:
```
Comment (Ebaneck) [inline docs/developer/best_practices/commit.rst]: This has several advantages amongst which are: ??
or
This has several advantages including: ??

**Review by alexandre-allard — DISMISSED** [review]

File: docs/developer/best_practices/commit.rst [hunk 25]
```diff
@@ -0,0 +1,372 @@
...
+Unlike a comment in the code (which applies to a single line or file), a commit
+message applies to a logical change and thus can provide information on the
+design of the code and why the change was done. This makes commit messages a
+part of the code documentation and makes them helpful for other developers to
+understand your code.
+
+Last but not least: commit messages can also be used for automating tasks such
+as issue management.
+
+Note that it is important to have all the necessary information in the commit
```
Comment (Ebaneck) [inline docs/developer/best_practices/commit.rst:100]: We could wrap the `note` in an `rst` note format since it is information you want the readers to retain always

For example ```..Note: ```

File: docs/developer/best_practices/commit.rst [hunk 26]
```diff
@@ -0,0 +1,372 @@
...
+    [… commit message body …]
+
+    Refs: #XXXXX
+    Refs: #YYYYY
+    Closes: #ZZZZZ
+    See: https://github.com/scality/metalk8s/issues/XXXXX
+    See: https://github.com/scality/metalk8s/issues/YYYYY
+    See: https://github.com/scality/metalk8s/issues/ZZZZZ
+
+Footer can also contains signature (``git commit -s``) or cherry-pick source
```
Comment (Ebaneck) [inline docs/developer/best_practices/commit.rst]: Footer can also contain a signature??

File: docs/developer/best_practices/commit.rst [hunk 27]
```diff
@@ -0,0 +1,372 @@
...
+
+- ``Bump Create React App to v3 and add optional-chaining``: Why? What are the
+  benefits?
+
+- ``Add skopeo & m2crypto to packages list``: Why do we need them?
+
+- ``Split certificates bootstrap between CA and clients``: Why do we need this
+  split? What is the issue we are trying to solve here?
+
+Note that none of these commits contain a reference to an issue (which
```
Comment (Ebaneck) [inline docs/developer/best_practices/commit.rst:253]: Ditto, maybe an rst note format here??

**Review by Ebaneck — CHANGES_REQUESTED** [review]
LGTM... just a couple of minors while reading through

File: docs/developer/best_practices/commit.rst (hunk 25 above)
Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst:100]: `Note that` here is more an expression than really introducing a note (I could totally remove it and just start the sentence with `It is important…`), si I'll keep it like that.

File: docs/developer/best_practices/commit.rst (hunk 24 above)
Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst]: Ditto

File: docs/developer/best_practices/commit.rst (hunk 27 above)
Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst:253]: Ditto

File: docs/developer/best_practices/commit.rst (hunk 26 above)
Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst]: Yeah they could.
Most of my commits end with `Signed-off-by: Sylvain Laperche <…> `

And actually, it's even required for contributors outside of Scality: https://github.com/scality/metalk8s/blame/development/2.4/CONTRIBUTING.md#L110-L121

File: docs/developer/best_practices/commit.rst (hunk 24 above)
Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst]: I'll use `This has several advantages amongst which are:`, thanks!

**Review by carlito-scality — DISMISSED** [review]

**Review by Ebaneck — DISMISSED** [review]

File: docs/developer/best_practices/commit.rst (hunk 26 above)
Comment (Ebaneck) [reply docs/developer/best_practices/commit.rst]: Yes yes...
Here, it was more of an issue with
`s/contains/contain a/`

Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst]: Oh right. Done.

**Review by MonPote — DISMISSED** [review]

File: docs/developer/best_practices/commit.rst [hunk 28]
```diff
@@ -0,0 +1,372 @@
...
+
+- make ``git log`` and ``git blame`` way more useful
+
+Examples
+^^^^^^^^
+
+The golden rule to create good commits is to ensure that the commit is atomic.
+A commit is atomic when:
+
+- it can be reverted without breaking the product
```
Comment (gdemonet) [inline docs/developer/best_practices/commit.rst]: Maybe that's a property we *want*, but it's definitely harder to achieve than having the ability to `checkout` any commit and have working code.

To me, `revert` ing a commit is a dangerous operation, and will always be: you never know what code was introduced after this commit, and if it relies on the contents you're removing. We are not reverting code often, and when we do, it's often one of two cases:
- disabling a commit in forward-merges (e.g. some implementation only needed for K8s versions before 1.15)
- removing a commit that works around a limitation of a dependency of our product, that we want to update

IMHO, `revert` is rarely used to remove an error unvoluntarily introduced. We have tests and reviews to prevent bugs from landing in the code, and shouldn't optimize our workflows to "patch" our mistakes after the act.

So, atomicity may be defined in terms of this ability to `revert`, but I'd rather focus on having the ability to **use** the code in any given point of history rather than being able to remove parts of it (not that usual of an operation). Furthermore, when `bisect`ing, ability to `revert` is irrelevant: the most important is that the code should not be broken at all, and we're trying to find one commit that is actually breaking something (maybe due to some history manipulation, or considering another, newer, view of _what is a correct state_).

Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst]: > use the code in any given point of history 

This seems pretty hard to achieve, because most of the code we add depends on already existing code, no?
So if some required code is renamed or moved, then past that point you won't be able to "use it" as-is.

---

Side note: I don't think we will ever come up with a precise definition for an atomic (reversible, logical, you name it) commit that'll stand a reality check (and that's why I didn't try to define it in my first version :stuck_out_tongue: ).

I think it's intrinsically ambiguous and everyone will see it a bit different.
Is it a big deal?
I'm not sure. Most of our current problem are regarding commit message, split is mostly OK.

/me currently wondering if he should go back to an intendedly open "definition"

Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst]: For the record, the concept of "logical commit" (without more definition) is coming from [here](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes), which was one of the three main sources that have influenced me.

Comment (gdemonet) [reply docs/developer/best_practices/commit.rst]: > > use the code in any given point of history

> This seems pretty hard to achieve, because most of the code we add depends on already existing code, no?
> So if some required code is renamed or moved, then past that point you won't be able to "use it" as-is.

That's not what I meant. My point is, any **merged** commit should work as is, i.e. for any commit in `development/x.y`, I can ` git checkout` said commit and have working code.

This implies:
- reordering is a dangerous operation
- `cherry-pick`ing and `revert`ing are dangerous as well

---

I like the "open" definition from your link, that would suit me.

File: docs/developer/best_practices/commit.rst (hunk 21 above)
Comment (gdemonet) [reply docs/developer/best_practices/commit.rst:164]: I'd even argue those *topics* in a commit subject are already expensive to use: e.g. `examples/solution: ` costs 19 out of 50 characters, which makes meaningful subjects harder to write.

~I'm wondering what the use-case of *topics* really is. Are we trying to have easy to understand lists of commits, from `git log --oneline` and the likes? Or to have easy to filter lists of commits? If the latter, then having such information in the subject is unnecessary.~
OK, just read through and found the explanation.

File: docs/developer/best_practices/commit.rst [hunk 29]
```diff
@@ -0,0 +1,372 @@
...
+  why was it buggy? —, some cleanup, some optimization, …).
+  It is really important to describe the intent/motivation behind the changes.
+
+- What change did you make? Document what the original problem was and how it
+  is being fixed (can be omitted for short obvious patches).
+
+- Why did you make the change in that way and not in another (mention alternate
+  solutions considered but discarded, if any)?
+
+When writing your message you must consider that your reader does not know
```
Comment (gdemonet) [inline docs/developer/best_practices/commit.rst:197]: I'd somewhat reduce this statement: it's almost impossible to write _"noob-friendly"_ commits while carrying out meaningful information for e.g. knowledgeable reviewers, without writing full pages for each change.

Maybe we could go for something closer to "Try to write your message in an explicit manner, adding references when concepts would need explanation. If the reader comes from outside of the core developers, she needs to be able to navigate through to understand the change".

File: docs/developer/best_practices/commit.rst [hunk 30]
```diff
@@ -0,0 +1,372 @@
...
+anything about the code you have patched.
+
+You should also describe any limitations of the current code. This will avoid
+reviewer pointing them out, and also inform future people looking at the code
+which tradeoffs were made at the time.
+
+Lines should be wrapped at 72 characters.
```
Comment (gdemonet) [inline docs/developer/best_practices/commit.rst]:
```suggestion
Lines must be wrapped at 72 characters.
```

File: docs/developer/best_practices/commit.rst [hunk 31]
```diff
@@ -0,0 +1,372 @@
...
+  lot of back and forth, which makes hard to get the gist of it.
+
+- you need access to an external service to get the whole context, which goes
+  against one of biggest advantage of the distributed SCM (having all the
+  information you need offline, from your local copy of the repository).
+
+- migration from one tracking system to another will invalidate the
+  references/links to the issues.
+
+Anatomy of a good commit message
```
Comment (gdemonet) [inline docs/developer/best_practices/commit.rst:113]: Can we provide a template? Users could "install" it locally, with `git config commit.template` :)

File: docs/developer/best_practices/commit.rst [hunk 32]
```diff
@@ -0,0 +1,372 @@
...
+reviewer pointing them out, and also inform future people looking at the code
+which tradeoffs were made at the time.
+
+Lines should be wrapped at 72 characters.
+
+Footer
+::::::
+
+Use references such as ``Refs``, ``See``, ``Fixes`` or ``Closes`` followed by
+an issue number to automate issue management.
```
Comment (gdemonet) [inline docs/developer/best_practices/commit.rst:211]: Such references can have a special meaning, let's mention it here: https://help.github.com/en/github/managing-your-work-on-github/closing-issues-using-keywords

**Review by cmonfort — DISMISSED** [review]

File: docs/developer/best_practices/commit.rst (hunk 32 above)
Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst:211]: Yeah, that's the how they "automate issue management", but I'll add the link, you're right.

File: docs/developer/best_practices/commit.rst (hunk 31 above)
Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst:113]: I don't have one ready right now.

Also, I'm worried it would be more of a hindrance than helpful since you'll always have text to delete (because you can hardly generate a useful default).

File: docs/developer/best_practices/commit.rst (hunk 30 above)
Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst]: Done

File: docs/developer/best_practices/commit.rst (hunk 29 above)
Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst:197]: It's not about noob-friendliness, as in "explain acronyms such as IPIP" or "what is a Pod?".
It's more about anyone should be able to review the code, even if it's not in their comfort zone.

That kinda goes hand in hand with the bullet points above: give context and explain why such as an outsider understand the commit.

Now, if we have a more clear way than "doesn't know anything about the code patched", I can reformulate :slightly_smiling_face:

File: docs/developer/best_practices/commit.rst (hunk 21 above)
Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst:164]: Yeah, topic can be quite a mouthful as well, but in you example I would replace `examples/solution` by `solution` since it stays unambiguous.

The "heuristic" I give (take the name of the top-level directory) is not perfect, it's more like an hint that works is most cases, but is not always optimal.

File: docs/developer/best_practices/commit.rst (hunk 28 above)
Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst]: > That's not what I meant. My point is, any merged commit should work as is, i.e. for any commit in development/x.y, I can git checkout said commit and have working code.

Ok, I agree with that.

> I like the "open" definition from your link, that would suit me.

Then I'll revert back to my "open" denomination. And if the need for a more precise definition occurs, we can then edit this document (it's not set in stone).

File: docs/developer/best_practices/commit.rst (hunk 21 above)
Comment (gdemonet) [reply docs/developer/best_practices/commit.rst:164]: There's a slight difference between `examples/solution` and `salt/solutions` though. But yes, let's do our best :)

File: docs/developer/best_practices/commit.rst (hunk 31 above)
Comment (gdemonet) [reply docs/developer/best_practices/commit.rst:113]: We can use comments for non-required content, I guess.

Something like:

```
topic: subject (stay below 50 chars)

# Commit message body
# Explains:
#   - why the change,
#   - what was changed,
#   - discarded alternatives

# Commit message footer
# References:
#   - issues fixed by this change, using e.g. `Fixes: #XXXXX`
#   - issues linked to this change, using e.g. `Refs: #XXXXX`
#   - external links, using e.g. `See: kubernetes/kubernetes/issues/XXXXX`
```

**Review by gdemonet — DISMISSED** [review]

File: docs/developer/best_practices/commit.rst (hunk 21 above)
Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst:164]: Yeah, we could interpret `solutions` as the `solutions` itself, and `salt/solutions` as the Salt code to deploy the solution.
But yeah, no hard rules (that's why I haven't listed a closed-set of accepted topic: the goal is mainly informative).

File: docs/developer/best_practices/commit.rst (hunk 31 above)
Comment (slaperche-scality) [reply docs/developer/best_practices/commit.rst:113]: Where would that template lives?
- Inline in this doc?
- Or somewhere in the repo?

(still not really convinced of the help of this template though ^^)

**Review by gdemonet — APPROVED** [review]
